_posts/18-05-2022-a_story_unfolds.md

```diff
@@ -0,0 +1,223 @@
+```haskell    
+tree :: File -> IO ()
+tree file = tree' "" file
+  where tree' prefix file = do
+          putStrLn (prefix ++ "- " ++ name file)
+          isDirectory <- isDir file
+          if isDirectory
+          then do
+            contents <- listDir file
+            traverse_ (tree' (' ':prefix)) contents
+          else return ()
+```
+It does not produce nice branch-looking lines between filenames, relying on indentation instead - but it is close enough.
+
+Obviously, this code is clumsy - and not just because the utility functions are crude. No, this code is just clumsy and inflexible in itself - and what's even worse, it makes one cardinal sin.
+_It mixes the responsibilities._
+
+While it is not the gravest sin of them all, as one could argue that _traversing_ the filesystem and _printing_ the filenames are very strongly related, it would be nice to do these two separately. In fact, the added flexibility might come in handy if we want to do something more than just printing - and, as we're about to see, at pretty much no added cost.
+
+So into the forest we go, preparing our tree-like data structure:
+```haskell
+data FileTree = RegularFile File | Directory File [FileTree]
+```
+Now we can write a function that grows these trees:
+```haskell
+buildTree :: File -> IO FileTree
+buildTree file = do
+  isDirectory <- isDir file
+  if isDirectory
+  then do
+    contents <- listDir file
+    subtrees <- traverse buildTree contents
+    return (Directory file subtrees)
+  else return (RegularFile file)
+```
+And we can restore the printing capability by writing another simple function:
+```haskell
+printTree :: FileTree -> String
+printTree = unlines . printTree'
+  where printTree' (RegularFile file) = ['-':(name file)]
+        printTree' (Directory file contents) = ('-':(name file)):(subtrees contents)
+        subtrees = map prepend . concatMap printTree'
+        prepend path = ' ':path
+```
+Reasonably straightforward. We could even get away with making `FileTree` derive `Show` typeclass. The main benefit, though, is that we have separated the process of traversing the filesystem from the process of listing its contents - and should we ever need a function that goes into a directory and removes everything below a certain depth, we already have half of it implemented. Moreover, if we want to recover our initial function, we just need to compose a few blocks:
+```haskell
+tree :: File -> IO ()
+tree file = do
+  dirTree <- buildTree file
+  let formatted = printTree dirTree
+  putStrLn formatted
+```
+But can we go even further?
+
+There seems to be nothing left to extract - we have separated traversal and printing, so what more could we possibly do? But there _is_ in fact a subtle thing. We are walking along the system paths as the trees unfold before us, but we do it in distinct steps: first we take a look around to gather all the possible paths; and then we walk along them.
```

Comment:
Shouldn't it say:
```suggestion
There seems to be nothing left to extract - we have separated traversal and printing, so what more could we possibly do? But there _is_ in fact a subtle thing. We are walking along the system paths as the trees unfold before us, but we COULD do it in distinct steps: first we take a look around to gather all the possible paths; and then we walk along them.
```

As in, the above impl does NOT do it in distinct steps, but now we'll hew it so that it does

Reply:
It was intentional - even in this implementation there are two distinct steps, connected only by the fact that one needs the output of the other. Anyway I probably should rephrase it a bit.